feat: Add the 2026 Debugging survey report (#407)
rust-lang/surveys: surveys/2026/debugging/report/blog.md

+++

One of the biggest challenges Rust developers report in our
[annual surveys][state-of-rust-2025] is a subpar debugging experience. So, back
in February, we ran our first [Rust Debugging Survey][rust-debugging-survey],
in the hopes of identifying how Rust developers are using debuggers and what
problems they are facing when doing so. We received over 2,300 responses, and
we'd like to thank everyone who took the time to participate in the survey!

In this report, we'll go over some of the results of the survey. If you'd like,
you can also check out [the complete results of the survey][report].

If you'd like to skip ahead to any particular section, you can do so with this
index:

- [Cohort Analysis](#cohort-analysis)
- [Debugger Usage](#debugger-usage)
- [Pain Points](#pain-points)
- [Library Authors](#library-authors)
- [Closing Remarks](#closing-remarks)

## Cohort Analysis

The first step to making sense of the survey results is understanding who took
the survey. We asked respondents to rate their Rust expertise, from "Never used
it" to "Advanced". Over 80% reported themselves as "Advanced" or "Intermediate",
split roughly evenly between the two:

> <!-- chart: how-would-you-rate-your-rust-expertise -->

Reviewer: I feel like these charts could become smaller, they don't contain that much information, but are currently pretty huge. `height=600` seems to do the trick.

```suggestion
<!-- chart: how-would-you-rate-your-rust-expertise (height=600) -->
```

We also asked respondents if they currently use or have used debuggers in Rust.
Over 46% said they currently do, with the remaining responses split between

> "have in the past" and "never have":

Reviewer: ```suggestion
We also asked respondents if they currently use or have used debuggers in Rust.
Over 46% said they currently do, with the remaining responses split between
"have in the past" and "never have". That means that over half of respondents do not currently use a debugger for Rust!
```

<!-- chart: do-you-use-debuggers-in-rust -->

Categorized by expertise, the responses reveal that roughly half of "beginners"
have never used debuggers in Rust, and nearly half of "advanced users" currently

> do use debuggers in Rust:

Reviewer: ```suggestion
Categorized by expertise, the responses reveal that roughly half of "beginners"
have never used debuggers in Rust, and advanced users are more likely to use a debugger in Rust:
```

Author: I'm hesitant to make this change because I feel the new wording may be implying a relationship that isn't necessarily true (something along the lines of "advanced users have the knowledge needed to benefit from debuggers", when the data could also support something like "using debuggers helps users learn enough about their code to consider themselves advanced users").

That, and I also don't want to reduce the amount of information directly available in the text if I can help it due to my accessibility concerns.

Reviewer: I just found the original text confusing, because it says "half of beginners do X and half of advanced users do Y", it switches both the target group and X -> Y in the same sentence, which just makes it confusing for me to read.

Author: Hm, that is a good point. Would breaking it up a bit help, maybe? Something like...

> Categorized by expertise, the responses reveal that roughly half of "beginners" have never used debuggers in Rust! On the other hand, nearly half of "advanced users" currently do use debuggers in Rust:

Reviewer: :+1:

> <!-- chart: do-you-use-debuggers-in-rust-per-expertise -->

Reviewer: This chart is a bit clipped:

<img width="915" height="519" alt="Image" src="https://github.com/user-attachments/assets/1696df84-0a83-4c71-b40c-2359174c0b6e" />

Adding some left margin should help.

For respondents who indicated they had previously used Rust but no longer did,
we asked if challenges with debugging support were why they stopped. For nearly
3%, the answer was "yes", with an additional 24% reporting debugging issues as
being partially responsible (though mind the small response count; most
respondents were active users of Rust):

<!-- chart: were-issues-with-debugging-support-the-primary-reason-why-you-stopped-using-rust -->

> ## Debugger Usage

Reviewer: I feel like this section should contain the "Do you use a debugger" chart? It seems kinda obvious that this question should belong here. Maybe just keep the cohort analysis and why people stopped using debuggers in the first section?

Author: The way I'm looking at it, "do you use a debugger" is more of a "who" question than a "how" or "why" question, particularly in the context of a survey all about debuggers. This is also part of why this question was the third question in the "Cohort Questions" portion of the actual survey.

Does considering that angle change your mind? If not, I'm happy to move things around as you suggested. I don't think it makes a huge difference either way.

Also, perhaps you misspoke, but:

> Maybe just keep the cohort analysis and why people stopped using debuggers in the first section?

That last question in the cohort analysis section is currently if debuggers were the reason people stopped using Rust, not why they stopped using debuggers.

Reviewer: Yeah sorry, I meant Rust, not debuggers :)

I'm fine with keeping the current structure in that case, but I don't like the "Cohort analysis" section name very much. Maybe we can directly base the section names based on what you write, and have "Who uses debuggers?" and "How are debuggers used?" or something like that?

Author: If you like those section names better, that's fine by me.

Knowing what debuggers developers are using and how is another important part of
understanding the challenges they face. To this end, we asked respondents how
they were debugging their programs on different operating systems. Independent
of operating system, most developers make use of print debugging and the `dbg!`
macro. Excluding those, using `lldb` inside an IDE was the most popular choice,

> followed by `gdb` on the command line:

Reviewer: ```suggestion
Knowing what debuggers developers are using and how is another important part of
understanding the challenges they face. To this end, we asked respondents how
they were debugging their programs. Unsurprisingly, most developers make use of print debugging and the `dbg!`
macro. Excluding those, using the `lldb` debugger inside an IDE was the most popular choice,
followed by `gdb` on the command line:
```

<!-- chart: what-tools-and-workflows-do-you-use-to-debug-rust-programs -->

Including the operating system as part of the response, we can get a more

> detailed breakdown of these results from two different angles. The first angle

Reviewer: ```suggestion
We can get a more detailed breakdown of these results if we include the operating system on which the respondents use a given debugging approach. We examine this from two different angles. The first angle
```

being, "On operating system X, what percent of responses are using debugger
Y?". Print debugging and the `dbg!` macro are consistently the top two yet
again, but looking beyond that, things get more interesting. On Linux, using
`gdb` on the command line was the most popular choice by a thin margin, but on
Windows, Windows Subsystem for Linux (WSL), and macOS, `lldb` in an IDE was the
top pick with a much wider lead over the respective runner-up for each compared

> to Linux (which had `lldb` in an IDE as the runner-up). On Windows, the three

Reviewer: This sentence is complicated to follow, I didn't understand the "for each" part.

Author: I will have to think a bit on how to reword this properly. For now, though, I will try to explain what I am going for.

> On Linux, using `gdb` on the command line was
the most popular choice by a thin margin,

With `gdb` CLI on Linux at 11.5% and `lldb` in an IDE on Linux at 11.1%, `gdb` only has a small lead over `lldb` (the thin margin).

> but on Windows, Windows Subsystem for
Linux (WSL), and macOS, `lldb` in an IDE was the top pick with a much wider lead
over the respective runner-up for each compared to Linux (which had `lldb` in an
IDE as the runner-up).

On Windows, `lldb` in an IDE leads with 13.3%, and is followed by the Visual Studio debugger at 7.27%.

On WSL, `lldb` in an IDE leads with 10.7%, and is followed by `gdb` CLI at 4.58%.

On macOS, `lldb` in an IDE leads with 15.5%, and is followed by `lldb` CLI at 8.03%.

| OS | "Top Pick" | "Runner-Up" | Difference |
|--------|--------|--------|--------|
| Linux | `gdb` CLI (11.5%) | `lldb` in IDE (11.1%) | ~0.4% |
| Windows | `lldb` in IDE (13.3%) | VS debugger (7.27%) | ~6.03% |
| WSL | `lldb` in IDE (10.7%) | `gdb` CLI (4.58%) | ~6.12% |
| macOS | `lldb` in IDE (15.5%) | `lldb` CLI (8.03%) | ~7.47% |

So when looking at the top two non-print-debugging debuggers per OS:

On Linux, even though the `gdb` CLI is preferred, the runner-up is not far behind. One could say that people could go "either way" between the two.

On Windows, WSL, and macOS, the preferred choice beat out the runner-up by at least 6%, making it a much more decisive victory. For all three, the preferred choice was `lldb` in an IDE, but they had different "runner-up" choices.

Reviewer: I think that this is a lot of nuance that I'm not sure if we even have to dig into. Maybe if it provided some super interesting insights, but I don't personally see this as being particularly interesting. Maybe we can just say that LLDB is popular across operating systems and people on Linux are much more likely to use a CLI than on other OSes?

Author: I think saying that using `lldb` in an IDE is a pretty popular option across operating systems would be reasonable. I'm not sure if saying people on Linux are much more likely to use a CLI is ideal, though. I can definitely see where that idea would come from, but I'm not sure it's the right wording. I think the second idea is more accurately represented with something like:

> Those on Linux-based and "Other" operating systems were more willing to work from the command line than others.

But that still feels like getting ahead of ourselves a bit. I already pointed out the reluctance to use CLI on Windows and the popularity of "IDK" on Windows and macOS, so maybe it's fine to just drop this part and focus on the popularity of `lldb` in an IDE. Something like...

> On Linux, using `gdb` on the command line was the most popular choice by a thin margin, beating `lldb` in an IDE by only 0.4%. On Windows, Windows Subsystem for Linux (WSL), and macOS, `lldb` in an IDE was the top pick by at least 6%, making it a very popular choice in general.

Reviewer: That looks good, feel free to use that.

least popular choices were the command line debuggers (`gdb` CLI, `lldb` CLI,
and BugStalker), and on both Windows and macOS the third most popular pick was,
"I don't know". Those who were debugging on operating systems not listed (Other)
most frequently used some kind of special embedded debugger or `gdb`:

<!-- chart: what-tools-and-workflows-do-you-use-to-debug-rust-programs-per-os-1 -->

The other angle we can look at these responses from is, "For users of debugger
X, what percent of responses are using it on operating system Y?". For most
debuggers, Linux makes up the largest portion of uses, ranging from about 45% to
about 77%, followed by Windows, then macOS. The most notable exceptions are
WinDbg and the Visual Studio debugger, which are primarily used on Windows, and
`lldb`, which is used more on macOS than Windows in an IDE and on the command
line:

<!-- chart: what-tools-and-workflows-do-you-use-to-debug-rust-programs-per-os-2 -->

As for how people actually use their debugger of choice, the aggregate results
are not particularly surprising. Roughly 87% of users are using debuggers for
stepping line-by-line through programs and a little over half of users are using

> debuggers to obtain stack traces from hung/crashed processes:

Reviewer: ```suggestion
To the 6 respondents who use WinDbg on Linux: we wish you luck!
```

Author: Am I understanding correctly that this suggested line should come after the chart it describes and replace the description for the chart that follows?

I do see that you've also suggested a new description for that following chart in https://github.com/rust-lang/surveys/pull/407#discussion_r3803817726, which would come after the chart. I'm not particularly against the suggested changes, but I worry that the exact placement of them could be a bit confusing. We usually have at least a sentence before each chart that "introduces it" (we did so for every chart in [the compiler performance survey report blog post]), it'd be a little odd for the only lines preceding this chart to be talking about the chart that came before it.

Also, the usual "accessibility concern has me worried about providing less information about the chart in text form".

[the compiler performance survey report blog post]: https://blog.rust-lang.org/2025/09/10/rust-compiler-performance-survey-2025-results/#overall-satisfaction

Reviewer: Sorry, I messed up the GitHub diff, I just wanted to add this as a new sentence, not remove the previous paragraph (the GitHub UI does not make it particularly easy to add new stuff at a specific place with a proposed diff). If you wan to keep the sentence, place it wherever you see fit :)

Author: Aaaah, gotcha. Added.

Author: This can be marked resolved if you're happy with the placement I chose in 5029165cfd9d501b05253bd1247bd3dbd50813dc.

<!-- chart: what-are-you-using-debuggers-for -->

> If we break these results down by expertise, we can learn a bit more about usage
> patterns. As users become more experienced with Rust, their use of debuggers for
> learning purposes decreases, and they get more stack traces from crashed
> processes:

Reviewer: ```suggestion
Only a quarter of the respondents use a debugger to debug async code. That might be partially caused by the async Rust debugging experience being clumsy and incomplete.

If we break these results down by expertise, we can learn a bit more about usage
patterns. As users become more experienced with Rust, their use of debuggers for
learning purposes decreases, but they spend more time analyzing stack traces from crashed
processes:
```

Author: Concerns bundled in https://github.com/rust-lang/surveys/pull/407#discussion_r3806697905

Reviewer: I might have also placed this in the wrong place.

Author: Does c855aa03f334030cfddef5ba57f0c604231ebfd4 look good?

Reviewer: Yup.

<!-- chart: what-are-you-using-debuggers-for-per-expertise -->

The final bit of insight into how Rustaceans use debuggers is if they are
debugging programs that use Rust alongside other programming languages. For 44%

> of respondents, the answer is "yes":

Reviewer: ```suggestion
The final bit of insight into how Rustaceans use debuggers is if they are
debugging programs that use Rust alongside other programming languages. For 44%
of respondents, the answer is "yes", which is a pretty high number!
```

<!-- chart: do-you-debug-programs-that-combine-rust-with-other-languages -->

As for which languages those are, C dominates the scene at a little over 70%,
followed by C++ at about 43% and Python at about 20%:

<!-- chart: do-you-debug-programs-that-combine-rust-with-any-of-the-following-languages -->

> ## Pain Points

Reviewer: ```suggestion
## Challenges
```

Instead of diving right into asking, "what problems do you face when using
debuggers?", or something to that effect, we first asked respondents why they
decide against using debuggers whenever they do, including for reasons that
aren't necessarily "problems with debuggers". This can help to put the severity
of any issues into context; if most users report they are blocked by poor
support for language features and very few report being blocked by "non-issues",

> it'd show the lack of support to be a very significant blocker for users.

Reviewer: I didn't understand the second half of this sentence - what are the "non-issues"? Could you maybe reword this paragraph a bit?

Author: I will ponder the best way to reword this, but for now, I will try to explain what I was going for.

The way I see it, some of these responses are clear indicators of a problem (i.e. "language features aren't supported") and others are much more likely to be either "not a problem" or "not *our* problem" (i.e. "my code works"). I think you could probably argue any of the responses could go either way (especially because of how we worded them), but I also think many of them "lean" a certain way (i.e., "I don't know how" reads to me as leaning towards "not a problem" as the user just needs to take the time to learn how, but you could certainly argue that this is caused by poor documentation, a UI/UX issue, or something else that makes it more of "our problem").

If 90% of people report they decide against using a debugger because their code works and 30% report they decide against using a debugger because language features aren't supported, that's concerning, but not the end of the world.

If we flip those numbers, then not only are we saying that 90% of people decided against using a debugger because language features aren't supported, but we're also saying that *only* 30% of them have ever decided against using a debugger because their code worked. I feel like that would suggest something about the severity of the lack of language support.

I might be trying to create meaning where there is none, though. Maybe it'd be better to drop this part?

Reviewer: I like the way you think deeply about these things :) I think that mentioning the high-level split is fine ("problems caused by poor debugger support" vs "I wouldn't use a debugger even if it was perfect"), but the rest is too complicated, so I would drop it, yeah.

Author: Dropped in cde4081ca74a6fb4b87553259d2dd95aebc4fc6e, I didn't add anything new since I feel the existing text already points out the high-level split. Let me know what you think?

Reviewer: Looks good :+1:

The most commonly reported reason was that it was easier or faster to use logs
or print debugging to solve problems, reported by a little over 81% of
respondents. This isn't all that surprising; one would expect a lot of small
simple problems just don't really need a debugger. It does leave one wondering
if the user experience could be made convenient enough to dethrone print
debugging, but it seems hard to beat something so intuitive. This is followed by
roughly 37% of respondents who write code that Just Works. Fair enough. After
that, about 26% of respondents indicated that they've decided not to use
debuggers in situations where the language features they were working with had
poor support. This is slightly more than issues with standard library types, at
about 22%, which is slightly more than issues with external library types, at
about 20%:

<!-- chart: when-you-dont-use-a-debugger-why-dont-you -->

As stepping through code was anticipated to be one of the most common uses for
debuggers, we directly asked respondents if they faced any issues when doing so.
A little over 51% of respondents said they did:

> <!-- chart: do-you-experience-any-issues-when-trying-to-step-through-code-with-your-debugger -->

Reviewer: This doesn't really need a chart, it's just two values, it takes space unnecessarily.

Author: Removed. I left the text, does that seem better?

Of those who reported that they experienced issues stepping through code, we
asked when they were experiencing issues. Async code was the most common case
reported at slightly over 28%, followed by code involving macros at about 23%.
The least common case reported was code involving function pointers, at almost
6%:

<!-- chart: when-do-you-experience-issues-with-trying-to-step-through-code-with-your-debugger -->

We also directly asked respondents which types in the standard library were hard
to work with, if any. This was an open-response question, and as can be seen in
the wordcloud, a few common complaints were with `enum`s and collections,
particularly `std::collections::HashMap` and `std::vec::Vec`:

<!-- TODO: This doesn't work! Wordclouds really weren't designed to exist without a
non-wordcloud chart. -->

> `chart: what-standard-library-types-are-hard-to-work-with-when-debugging-wordcloud`

Reviewer: I don't think we need to show the wordcloud, we can just describe it with a sentence.

As the survey was being created, we interviewed a few community members and did
some brainstorming to think of a handful of pain points people might experience
when using debuggers with Rust. We asked respondents to indicate which, if any,
they have encountered. At slightly over 74%, poor representation of values was
the most common pain point by a decent margin, followed by being unable to print
variables at just over 55%:

<!-- chart: which-of-these-pain-points-have-you-experienced-using-a-debugger-with-rust -->

> ## Library Authors

Reviewer: ```suggestion
## Debugger visualizers
```

We asked respondents to indicate if they were library authors, and if so, if
they were aware of and using the `debugger_visualizer` attribute. Nearly 62% of
respondents indicated that they were library authors who were not aware of this
attribute:

> <!-- TODO: Fix this chart in the script, then consider rewriting this section. -->

Reviewer: This chart is also kinda useless, the 62% number if enough.

Author: Really? Is that based on the chart as it currently exists (broken!) or as it should exist? Because it *should* also show the percent who meet the criteria "library authors who knew about the attribute but did not use it", but it currently doesn't.

That criteria is directly relevant to the next chart, which was only offered to those individuals.

I feel like the chart would be more useful to have if it correctly displayed the four distinct categories.

Reviewer: Oh, sure, if the chart shows more information, then feel free to keep it. I just think that it is pretty much useless to have a gigantic pie chart that shows two categories. 3+ is fine.

Author: Yeah, it was showing as two categories because we were normalizing the answers to the question in one place but we weren't in another place, leading to the three answers ending in periods being unknown and treated as "Other". I managed to fix this with a030ce9a657a18a4d9f5661d98c47d299259013f, though I'm not 100% sure if that's the right long-term solution or not. It *seems* fine to me.
